### ISSUE TITLE
Overhaul serialization

### ISSUE THREAD

**Issue by Ambrevar** [atlas-engineer/nyxt#2644]
See  https://github.com/atlas-engineer/nfiles/issues/6#issuecomment-1239497700 for an initial discussion.

The libraries that serialize Lisp to human-readable text:

- cl-prevalence: https://github.com/40ants/cl-prevalence
- marshal: https://github.com/wlbr/cl-marshal
- cl-store: https://github.com/skypher/cl-store/

They all suffer from a few issues, namely around formatting.  See for instance https://github.com/40ants/cl-prevalence/issues/created_by/ambrevar.

The API of at least cl-prevalence is a bit clunky.  I'm working on it (see https://github.com/40ants/cl-prevalence/pull/21 for instance) but some of it won't ever be fixed (like the impedance mismatch between the s-exp and the XML backends).

Finally, performance is almost certainly subpar, which for the GHT can incur a performance hit.

In https://github.com/atlas-engineer/nfiles/issues/6#issuecomment-1239497700 we discussed leveraging `make-load-form` which is a standard function:

- benefits: built-in, high performance. `make-load-form-saving-slots` is convenient.
- drawback: not human-readable.  Only `make-load-form-saving-slots`, no other convenience function by default,
  for instance to customize the serialization of a specific slot.

Another thing that could be neat would be to leverage MOP to customize how slots are serialized.  Indeed, it's pretty clunky to customize slot serialization at the moment with cl-prevalence.  It's going to get better with https://github.com/40ants/cl-prevalence/pull/21.
With MOP, we could add new slot options:

- `:load-form :initform` (thus not serialized)
- `:load-form` :current-value (the default)
- `:load-form (lambda (object slot-name) (make-instance ... ))`

Is it really better than https://github.com/40ants/cl-prevalence/pull/21 though?  Because lambdas are not as powerful as the generic functions as used in cl-prevalence (e.g. no `call-next-method`).

Thoughts? @aartaka

**Comment by aadcg** [comment]
@Ambrevar is 4.0 the right label?

**Comment by Ambrevar** [comment]
Yes, because changing serialization method would break the history and other persisted files.  It's mostly an implementation detail that helps the developers, it's not so important from a user perspective.

**Comment by Ambrevar** [comment]
I've put `high` because serialization happens all over the place.

**Comment by aartaka** [comment]
And in one place serialization is critical: history. If the history is malformed or reaches more than some 100-200 megabytes (not that many to never reach, I usually get to it in a month), `cl-prevalence`simply segfaults the whole image exhausting the memory. 

That's why we need to fix serialization before 3.0—we don't want Nyxt to segfault under the user in the middle of some important activity just because `cl-prevalence` exhausted all the memory.

**Comment by hendursaga** [comment]
As mentioned in #2435, using a SQLite backend as at least an alternative might be something to look into.. I don't really get the whole thing about saving history and bookmarks as S-exps - auto-mode rules, sure, that makes sense, but not something that's quite performance-critical and something with which interoperability would be a huge plus..

**Comment by Ambrevar** [comment]
Bookmarks are no problem at all, and they are super fast.  Human-readable and writable bookmarks are definitely a plus.

For the GHT, it's  more debatable, sure, but s-exp still have lots of benefits.

In any case, an SQLite database has no benefit in my understanding compared to a binary dump: it's more work + we would use no SQL feature at all beside persistence.  (The GHT does not map an SQL database by design.)

**Comment by aartaka** [comment]
@Ambrevar, GHT maps to relational DB model, and I bet I can do a DB scheme that would support it. But yeah, do we need it for anything that's not persistence and import/export?

**Comment by aartaka** [comment]
If it's only import/export, then SQL support is a matter of extension, not Nyxt core (。ヘ°)

**Comment by Ambrevar** [comment]
Agreed.

**Comment by aadcg** [comment]
GHT is history. See #3598.
